cpp/src/arrow/compute/kernels/isin.cc

```diff
@@ -0,0 +1,303 @@
+#include "arrow/util/macros.h"
+#include "arrow/util/string_view.h"
+#include "arrow/visitor_inline.h"
+
+namespace arrow {
+
+using internal::checked_cast;
+using internal::DictionaryTraits;
+using internal::HashTraits;
+
+namespace compute {
+
+class IsInKernelImpl : public UnaryKernel {
+  virtual Status Compute(FunctionContext* ctx, const ArrayData& right) = 0;
+
+ public:
+  Status Call(FunctionContext* ctx, const Datum& right, Datum* out) override {
+    DCHECK_EQ(Datum::ARRAY, right.kind());
+    const ArrayData& right_data = *right.array();
+    RETURN_NOT_OK(Compute(ctx, right_data));
+    return Status::OK();
+  }
+
+  std::shared_ptr<DataType> out_type() const override { return boolean(); }
+
+  virtual Status CompareArray(const Datum& left, Datum* out) = 0;
+};
+
+// ----------------------------------------------------------------------
+// Using a visitor create a memo_table_ for the right array
+// TODO: Implement for small lists
+
+template <typename Type, typename Scalar>
+struct MemoTableRight {
+  Status VisitNull() { return Status::OK(); }
+
+  Status VisitValue(const Scalar& value) {
+    memo_table_->GetOrInsert(value);
+    return Status::OK();
+  }
+
+  Status Append(const ArrayData& right) {
+    memo_table_.reset(new MemoTable(0));
+    return ArrayDataVisitor<Type>::Visit(right, this);
+  }
+
+  using MemoTable = typename HashTraits<Type>::MemoTableType;
+  std::unique_ptr<MemoTable> memo_table_;
+};
+
+// ----------------------------------------------------------------------
+// Check if value in both arrays or not and returns boolean values
+
+// \brief Iterate over the left array using another visitor.
+// In VisitValue, use the memo_table_ (for right array) and check if value
+// in left array is in the memo_table_. Append the true to the
+// BooleanBuilder if condition satisfied, else false.
+
+// Additional member "right_null_count" is used to check if
+// null count in right is not 0, then append true to the BooleanBuilder
+// when left array has a null, else append null.
+
+template <typename T, typename Scalar>
+class IsInKernel : public IsInKernelImpl {
+ public:
+  IsInKernel(const std::shared_ptr<DataType>& type, MemoryPool* pool) {}
+
+  Status VisitNull() {
+    if (right_null_count != 0) {
+      bool_builder_.UnsafeAppend(true);
+    } else {
+      bool_builder_.UnsafeAppendNull();
+    }
+    return Status::OK();
+  }
+
+  Status VisitValue(const Scalar& value) {
+    bool_builder_.UnsafeAppend(memo_table_->Get(value) != -1);
+    return Status::OK();
+  }
+
+  Status Compute(FunctionContext* ctx, const ArrayData& right) override {
+    memo_table_.reset(new MemoTable(0));
+
+    MemoTableRight<T, Scalar> func;
+    func.Append(right);
+    memo_table_ = std::move(func.memo_table_);
+
+    right_null_count = right.GetNullCount();
+    return Status::OK();
+  }
+
+  Status CompareArray(const Datum& left, Datum* out) override {
+    const ArrayData& left_data = *left.array();
+    bool_builder_.Reset();
```

Comment:
This might have conflicted with @fsaintjacques advice.  But I don't think we should not be allocating with Reserve with  a kernel when we know the exact length of the type (this is what PrimitiveAllocatingUnaryKernel is for)